cli/azd/cmd/root.go

```diff
@@ -121,35 +120,40 @@ type designBuilder[F any] func(opts *internal.GlobalCommandOptions) (*cobra.Comm
 type actionBuilder[F any] func(
 	console input.Console,
 	ctx context.Context,
+	chain []middleware.Middleware,
 	o *internal.GlobalCommandOptions,
 	flags F,
-	args []string) (actions.Action, error)
-
-type buildOptions struct {
-	disableTelemetry bool
-}
+	args []string) (actions.MiddlewareEnabledAction, error)
 
 func BuildCmd[F any](
 	opts *internal.GlobalCommandOptions,
 	buildDesign designBuilder[F],
 	buildAction actionBuilder[F],
-	buildOptions *buildOptions) *cobra.Command {
+	buildOptions *actions.BuildOptions) *cobra.Command {
 	cmd, flags := buildDesign(opts)
 	cmd.Flags().BoolP("help", "h", false, fmt.Sprintf("Gets help for %s.", cmd.Name()))
 
-	runCmd := func(cmd *cobra.Command, ctx context.Context, args []string) error {
+	cmd.RunE = func(cmd *cobra.Command, args []string) error {
+		ctx := cmd.Context()
+
 		console, err := initConsole(cmd, opts)
 		if err != nil {
 			return err
 		}
 
-		action, err := buildAction(console, ctx, opts, *flags, args)
+		middlewareChain := []middleware.Middleware{
+			middleware.NewDebugMiddleware(console),
+		}
+
+		if !buildOptions.DisableTelemetry {
+			middlewareChain = append(middlewareChain, middleware.NewTelemetryMiddleware(cmd.Name()))
+		}
+
+		action, err := buildAction(console, ctx, middlewareChain, opts, *flags, args)
```

Comment:
The concept of a middleware is a higher order operation than an action.  In this example the ideas/concept get muddled together the middleware is run as part of the action.

With regard to the telemetry, the middleware should be responsible to determining whether or not it should take any action, not the command function.

Reply:
> The concept of a middleware is a higher order operation than an action. In this example the ideas/concept get muddled together the middleware is run as part of the action.

That is indeed why we now introduce a composition of `middlewareRunner` that uses the command `action`. This is no different than your previous implementation, or the new one with runtime DI.

To be clear here, an `action` is simply an interface that looks like: `Run(ctx context.Context) (*ActionResult, error)`. In that regard, the middleware runner satisfies the same interface. Alternatively, we can change this to specifically target `middlewareRunner`, and then it's no different than what we had before.

> With regard to the telemetry, the middleware should be responsible to determining whether or not it should take any action, not the command function.

Would be interested to know about the "why" here to prefer always running telemetry middleware with a bool to skip logic. I think optional registration of middleware is perfectly fine as a concept, and simplifies runtime logic and makes it cleaner in understanding what should/shouldn't be running.